### PR TITLE
[CMS-629] Improve libraries handling by checking packagist.

### PR DESCRIPTION
**Pull request by kporras07**

### CODE REVIEW

File: src/Commands/ConvertToComposerSiteCommand.php [hunk 1]
```diff
@@ -112,9 +112,12 @@ public function convert(
             $sourceComposerJson,
             $this->getLocalSitePath(),
             $contribProjects,
-            $libraryProjects
+            $libraryProjects,
+            $this->drupalProjects->getSiteRootPath() . '/libraries-backup'
```
Comment (jabxjab) [inline src/Commands/ConvertToComposerSiteCommand.php:116]:
```php
Files::buildPath($this->drupalProjects->getSiteRootPath(), 'libraries-backup`);
```

File: src/Utils/DrupalProjects.php [hunk 2]
```diff
@@ -123,12 +137,44 @@ public function getLibraries(): array
...
         return $packages;
     }
 
+    /**
+     * Backup given library folder.
+     */
+    private function backupLibrary(string $filePath): void
+    {
+        $filesystem = new Filesystem();
+        $backupPath = Files::buildPath($this->siteRootPath, 'libraries-backup');
```
Comment (greg-1-anderson) [inline src/Utils/DrupalProjects.php:171]: Do non-Composer libraries have to go in a different path? Could we put them in `libraries`, and add a `!path/libraries/name` entry to .gitignore to allow each one to be committed to the repository on a case-by-case basis?

Comment (kporras07) [reply src/Utils/DrupalProjects.php:171]: I don't think they should go to a different path. I think updating .gitignore is a good idea but I thought that would be done as part of CMS-630 so, right now I'm just force-adding the libraries. Let me know if you prefer to do that here

Comment (kporras07) [reply src/Utils/DrupalProjects.php:171]: Re-reading this. They're going there only as a backup during the conversion process. They're restored later to the normal /libraries path

Comment (greg-1-anderson) [reply src/Utils/DrupalProjects.php:171]: Let's block this PR on CMS-630, then, and decide after that code is ready whether we want to use it to update the .gitignore here, or just modify .gitignore with code local to this function.

The part about the libraries being temporarily moved to `libraries-backup` and then moved to their correct location later is fine; I just missed the later step last time.

File: src/Utils/Git.php [hunk 3]
```diff
@@ -82,17 +82,25 @@ public static function getGitignoreDiff(string $baseFilePath, string $fileToComp
      *   The commit message.
      * @param null|array $files
      *   The files to stage.
+     * @param null|bool $force
+     *   Force adding changes (use -f flag).
      *
      * @throws \Pantheon\TerminusConversionTools\Exceptions\Git\GitException
      */
-    public function commit(string $commitMessage, ?array $files = null): void
+    public function commit(string $commitMessage, ?array $files = null, ?bool $force = false): void
```
Comment (jabxjab) [inline src/Utils/Git.php:90]: No reason for the `$force` to accept `null`

File: src/Utils/Git.php [hunk 4]
```diff
@@ -82,17 +82,25 @@ public static function getGitignoreDiff(string $baseFilePath, string $fileToComp
...
     {
+        $addArguments = [];
```
Comment (jabxjab) [inline src/Utils/Git.php:92]: Could be optimized in a way like:
```php
$command = null === $files ? ['add', '-A'] : ['add', ...$files];

if ($force) {
    $command[] = '-f';
}

$this->execute($addArguments);
```

File: src/Commands/Traits/MigrateComposerJsonTrait.php [hunk 5]
```diff
@@ -61,6 +72,27 @@ protected function migrateComposerJson(
...
 
         $this->copyComposerPackagesConfiguration();
+
+        if ($librariesBackupPath && is_dir($librariesBackupPath)) {
+            $this->restoreLibraries($librariesBackupPath);
+        }
+    }
+
+    /**
+     * Restore libraries from the backup path and commit them.
```
Comment (jabxjab) [inline src/Commands/Traits/MigrateComposerJsonTrait.php:82]: Missing `@param`

File: src/Commands/Traits/MigrateComposerJsonTrait.php [hunk 6]
```diff
@@ -61,6 +72,27 @@ protected function migrateComposerJson(
...
+     */
+    private function restoreLibraries(string $librariesBackupPath): void
+    {
+        $finder = new Finder();
+        $filesystem = new Filesystem();
+        foreach ($finder->directories()->in($librariesBackupPath)->depth(0) as $folder) {
+            $filesystem->mirror($folder->getPathname(), $this->localSitePath . '/web/libraries/' . $folder->getRelativePathname());
```
Comment (jabxjab) [inline src/Commands/Traits/MigrateComposerJsonTrait.php:89]: Use `Files::buildPath()`

File: src/Commands/Traits/MigrateComposerJsonTrait.php [hunk 7]
```diff
@@ -61,6 +72,27 @@ protected function migrateComposerJson(
...
+            $this->getGit()->commit('Copy library ' . $folder->getRelativePathname(), [$this->localSitePath . '/web/libraries/' . $folder->getRelativePathname()], true);
```
Comment (jabxjab) [inline src/Commands/Traits/MigrateComposerJsonTrait.php:90]: Use `Files::buildPath()`

File: src/Utils/DrupalProjects.php [hunk 8]
```diff
@@ -112,6 +124,8 @@ public function getLibraries(): array
             }
         }
 
+        $packagistBaseUrl = 'https://packagist.org/packages/{package}.json';
```
Comment (jabxjab) [inline src/Utils/DrupalProjects.php:127]: Consider introducing a class constant (ie. `self::PACKAGIST_BASE_URL`)

File: src/Utils/DrupalProjects.php [hunk 9]
```diff
@@ -123,12 +137,44 @@ public function getLibraries(): array
...
+        }
+
+        $finder = new Finder();
+        foreach ($this->getLibrariesDirectories() as $librariesDir) {
+            $finder->directories()->in($librariesDir)->depth(0);
+            if (!$finder->hasResults()) {
+                continue;
+            }
+
+            foreach ($finder as $folder) {
```
Comment (jabxjab) [inline src/Utils/DrupalProjects.php:155]: Mixed usage of "dir"/"directory" vs "folder". My suggestion would be to stick with the established naming ("directory")

File: src/Utils/DrupalProjects.php (hunk 2 above)
Comment (kporras07) [reply src/Utils/DrupalProjects.php:171]: Done

**Review by greg-1-anderson — APPROVED** [review]

**Pull request merged**